File: csharp/Noise.cs
public class Noise : Benchmark
{
    private struct Vec2
    {
        public double X;
        public double Y;

        public Vec2(double x, double y) { X = x; Y = y; }
    }

    private class Noise2DContext
    {
        private readonly Vec2[] _rgradients;
        private readonly int[] _permutations;
        private readonly int _sizeMask;

        public Noise2DContext(int size)
        {
            _sizeMask = size - 1;
            _rgradients = new Vec2[size];
            _permutations = new int[size];

            for (int i = 0; i < size; i++)
            {
                double v = Helper.NextFloat() * Math.PI * 2.0;
                _rgradients[i] = new Vec2(Math.Cos(v), Math.Sin(v));
                _permutations[i] = i;
            }

            for (int i = 0; i < size; i++)
            {
                int a = Helper.NextInt(size);
                int b = Helper.NextInt(size);
                (_permutations[a], _permutations[b]) = (_permutations[b], _permutations[a]);
            }
        }

        private static double Gradient(in Vec2 orig, in Vec2 grad, in Vec2 p)
        {
            double spX = p.X - orig.X;
            double spY = p.Y - orig.Y;
            return grad.X * spX + grad.Y * spY;
        }

        private static double Lerp(double a, double b, double v) => a * (1.0 - v) + b * v;
        private static double Smooth(double v) => v * v * (3.0 - 2.0 * v);

        private Vec2 GetGradient(int x, int y)
        {
            int idx = _permutations[x & _sizeMask] + _permutations[y & _sizeMask];
            return _rgradients[idx & _sizeMask];
        }

        public double Get(double x, double y)
        {
            double x0f = Math.Floor(x);
            double y0f = Math.Floor(y);
            int x0 = (int)x0f;
            int y0 = (int)y0f;

            Vec2 g00 = GetGradient(x0, y0);
            Vec2 g10 = GetGradient(x0 + 1, y0);
            Vec2 g01 = GetGradient(x0, y0 + 1);
            Vec2 g11 = GetGradient(x0 + 1, y0 + 1);

            Vec2 p = new(x, y);

            double v0 = Gradient(new(x0f, y0f), g00, p);
            double v1 = Gradient(new(x0f + 1.0, y0f), g10, p);
            double v2 = Gradient(new(x0f, y0f + 1.0), g01, p);
            double v3 = Gradient(new(x0f + 1.0, y0f + 1.0), g11, p);

            double fx = Smooth(x - x0f);
            double vx0 = Lerp(v0, v1, fx);
            double vx1 = Lerp(v2, v3, fx);

            double fy = Smooth(y - y0f);
            return Lerp(vx0, vx1, fy);
        }
    }

    private static readonly char[] SYM = [' ', '░', '▒', '▓', '█', '█'];
    private const int SYM_LENGTH = 6;

    private long _size;
    private uint _result;
    private Noise2DContext _n2d;

    public Noise()
    {
        _result = 0;
        _size = ConfigVal("size");
        _n2d = new Noise2DContext((int)_size);
    }

    public override void Run(long IterationId)
    {
        double yAdd = IterationId * 128 * 0.1;
        double step = 0.1;

        for (long y = 0; y < _size; y++)
        {
            double yf = y * step + yAdd;
            for (long x = 0; x < _size; x++)
            {
                double xf = x * step;
                double v = _n2d.Get(xf, yf) * 0.5 + 0.5;

                int idx = (int)(v * 5.0);
                if (idx >= SYM_LENGTH) idx = SYM_LENGTH - 1;
                _result += (uint)SYM[idx];
            }
        }
    }

    public override uint Checksum => _result;
}

File: fsharp/Noise.fs
namespace Benchmarks

open System

[<Struct>]
type Vec2 = { X: float; Y: float }

[<AllowNullLiteral>]
type Noise2DContext(size: int) =
    let sizeMask = size - 1
    let rgradients = Array.init size (fun _ -> 
        let v = Helper.NextFloat(1.0) * Math.PI * 2.0
        { X = Math.Cos(v); Y = Math.Sin(v) })

    let permutations = 
        let arr = Array.init size id
        for i in 0..size-1 do
            let a = Helper.NextInt(size)
            let b = Helper.NextInt(size)
            let temp = arr.[a]
            arr.[a] <- arr.[b]
            arr.[b] <- temp
        arr

    static member inline gradient (orig: Vec2) (grad: Vec2) (p: Vec2) =
        let spX = p.X - orig.X
        let spY = p.Y - orig.Y
        grad.X * spX + grad.Y * spY

    static member inline lerp a b v = a * (1.0 - v) + b * v
    static member inline smooth v = v * v * (3.0 - 2.0 * v)

    member private this.getGradient x y =
        let idx = permutations.[x &&& sizeMask] + permutations.[y &&& sizeMask]
        rgradients.[idx &&& sizeMask]

    member this.Get(x: float, y: float) =
        let x0f = Math.Floor(x)
        let y0f = Math.Floor(y)
        let x0 = int x0f
        let y0 = int y0f
        let x1 = x0 + 1
        let y1 = y0 + 1

        let g00 = this.getGradient x0 y0
        let g10 = this.getGradient x1 y0
        let g01 = this.getGradient x0 y1
        let g11 = this.getGradient x1 y1

        let p = { X = x; Y = y }

        let v0 = Noise2DContext.gradient { X = x0f; Y = y0f } g00 p
        let v1 = Noise2DContext.gradient { X = x0f + 1.0; Y = y0f } g10 p
        let v2 = Noise2DContext.gradient { X = x0f; Y = y0f + 1.0 } g01 p
        let v3 = Noise2DContext.gradient { X = x0f + 1.0; Y = y0f + 1.0 } g11 p

        let fx = Noise2DContext.smooth (x - x0f)
        let vx0 = Noise2DContext.lerp v0 v1 fx
        let vx1 = Noise2DContext.lerp v2 v3 fx

        let fy = Noise2DContext.smooth (y - y0f)
        Noise2DContext.lerp vx0 vx1 fy

type Noise() =
    inherit Benchmark()

    [<Literal>]
    let symLength = 6
    let sym = [| ' '; '░'; '▒'; '▓'; '█'; '█' |]

    let mutable size = 0L
    let mutable result = 0u
    let mutable n2d : Noise2DContext = Unchecked.defaultof<Noise2DContext>

    override this.Checksum = result

    override this.Prepare() =
        size <- Helper.Config_i64("Noise", "size")
        result <- 0u
        n2d <- Noise2DContext(int size)

    override this.Run(IterationId: int64) =
        if not (obj.ReferenceEquals(n2d, null)) then
            let yAdd = float (IterationId * 128L) * 0.1
            let stepX = 0.1
            let stepY = 0.1

            let mutable yf = yAdd
            for y in 0L..size-1L do
                let mutable xf = 0.0
                for x in 0L..size-1L do
                    let v = n2d.Get(xf, yf) * 0.5 + 0.5

                    let idx = int (v * 5.0)
                    let idx' = if idx >= symLength then symLength - 1 else idx
                    result <- result + uint32 (int sym.[idx'])

                    xf <- xf + stepX
                yf <- yf + stepY